--- a/kernel/kallsyms.c
+++ b/kernel/kallsyms.c
@@ -662,7 +662,7 @@
 			   iter->type, iter->name);
 #else
 	{
-		if (strstr(iter->name, "ksu_") || !strncmp(iter->name, "susfs_", 6) || !strncmp(iter->name, "ksud", 4)) {
+		if (strstr(iter->name, "ksu_") || strstr(iter->name, "kernelsu") || !strncmp(iter->name, "susfs_", 6) || !strncmp(iter->name, "ksud", 4)) {
 			return 0;
 		}
 		seq_printf(m, "%px %c %s\n", value,
